### PR TITLE
Add support for building CoreCLR for iOS/tvOS simulator

### PR DESCRIPTION
**Pull request by filipnavara**
Build instructions:
- Build the runtime pack with `./build.sh clr+clr.nativeaotlibs+libs+packs -os [iossimulator/tvossimulator/maccatalyst] -arch arm64 -cross`
- Run the sample app in iOS simulator: `./dotnet.sh publish src/mono/sample/iOS/Program.csproj -c Release /p:TargetOS=iossimulator /p:TargetArchitecture=arm64 /p:DeployAndRun=true /p:UseMonoRuntime=false /p:RunAOTCompilation=false /p:MonoForceInterpreter=false`

SDK would need to be updated with RIDs in `KnownRuntimeFramework` to use this. Xamarin counterpart for bindings support: https://github.com/xamarin/xamarin-macios/compare/net8.0...filipnavara:xamarin-macios:net8.0-coreclr-ioslike?expand=1

To do:
- [x] Add documentation
- [x] Cleanup
- [x] Fix RegisterToggleRef in Xamarin
- [x] Check and fix x64 builds (src/native/minipal/cpufeatures.c is broken in cross-build from arm64 -> x64 because it tries to use target's assembler, and for some reason CLR_CROSS_COMPONENTS_BUILD doesn't trigger its exclusion)
- [ ] Fix running iossimulator-x64 sample on arm64 (AppleAppBuilder produces XCode project with wrong arch)
- [ ] Package the static `libcoreclr_static.a`
- [ ] Fix runtime test builds

### CODE REVIEW

**Comment by filipnavara** [comment]
Obligatory screenshot:
image

**Comment by filipnavara** [comment]
Unfortunately Objective-C marshalling support is still broken, which is blocker for MAUI:
```
*** Terminating app due to uncaught exception 'System.Reflection.TargetInvocationException', reason: 'Exception has been thrown by the target of an invocation. (System.Reflection.TargetInvocationException)
     at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean wrapExceptions)
     at Microsoft.Maui.Controls.DependencyService.Get[T](DependencyFetchTarget fetchTarget)
     at Microsoft.Maui.Controls.Application.<.ctor>b__7_0()
     at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
     at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
     at System.Lazy`1.CreateValue()
     at Microsoft.Maui.Controls.Application.get_SystemResources()
     at Microsoft.Maui.Controls.ResourcesExtensions.GetMergedResources(IElementDefinition element)
     at Microsoft.Maui.Controls.Element.SetParent(Element value)
     at Microsoft.Maui.Controls.Element.OnChildAdded(Element child)
     at Microsoft.Maui.Controls.Element.AddLogicalChild(Element element)
     at Microsoft.Maui.Controls.Application.AddWindow(Window window)
     at Microsoft.Maui.Controls.Application.Microsoft.Maui.IApplication.CreateWindow(IActivationState activationState)
     at Microsoft.Maui.Platform.ApplicationExtensions.CreatePlatformWindow(IApplication application, UIWindowScene windowScene, NSDictionary[] states)
     at Microsoft.Maui.Platform.ApplicationExtensions.CreatePlatformWindow(IUIApplicationDelegate platformApplication, IApplication application, UIApplication uiApplication, NSDictionary launchOptions)
     at Microsoft.Maui.MauiUIApplicationDelegate.FinishedLaunching(UIApplication application, NSDictionary launchOptions)
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
     at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
  --- End of stack trace from previous location ---
     at ObjCRuntime.Runtime.InvokeMethod(MethodBase method, Object instance, IntPtr native_parameters)
     at ObjCRuntime.Runtime.InvokeMethod(MonoObject* methodobj, MonoObject* instanceobj, IntPtr native_parameters)
     at ObjCRuntime.Runtime.bridge_runtime_invoke_method(MonoObject* method, MonoObject* instance, IntPtr parameters, IntPtr* exception_gchandle)
  ECall methods must be packaged into a system module. (System.Security.SecurityException)
     at Foundation.NSObject.RegisterToggleReference(NSObject obj, IntPtr handle, Boolean isCustomType)
     at Foundation.NSObject.MarkDirty(Boolean allowCustomTypes)
     at Foundation.NSObject.MarkDirty()
     at Foundation.NSNotificationCenter.AddObserverToList(NSObject observer, String aName, NSObject anObject)
     at Foundation.NSNotificationCenter.AddObserver(NSObject observer, Selector aSelector, NSString aName, NSObject anObject)
     at Foundation.NSNotificationCenter.AddObserver(NSString aName, Action`1 notify, NSObject fromObject)
     at Foundation.NSNotificationCenter.AddObserver(NSString aName, Action`1 notify)
     at UIKit.UIApplication.Notifications.ObserveContentSizeCategoryChanged(EventHandler`1 handler)
     at Microsoft.Maui.Controls.Compatibility.Platform.iOS.ResourcesProvider..ctor()
     at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean wrapExceptions)
```

**Comment by filipnavara** [comment]
...and MAUI:
image

File: src/coreclr/hosts/inc/coreclrhost.h [hunk 1]
```diff
@@ -19,9 +19,15 @@
 // For each hosting API, we define a function prototype and a function pointer
 // The prototype is useful for implicit linking against the dynamic coreclr
 // library and the pointer for explicit dynamic loading (dlopen, LoadLibrary)
+#ifdef __cplusplus
 #define CORECLR_HOSTING_API(function, ...) \
     extern "C" int CORECLR_CALLING_CONVENTION function(__VA_ARGS__); \
```
Comment (am11) [inline src/coreclr/hosts/inc/coreclrhost.h:24]: nit: there is `EXTERN_C` macro to avoid the condition.

Comment (filipnavara) [reply src/coreclr/hosts/inc/coreclrhost.h:24]: This header is meant for external consumption and it cannot have other dependencies.

File: src/tasks/AppleAppBuilder/AppleAppBuilder.csproj [hunk 2]
```diff
@@ -9,6 +9,7 @@
   </PropertyGroup>
   <ItemGroup>
     <EmbeddedResource Include="Templates\*.*" />
+    <EmbeddedResource Include="..\..\coreclr\hosts\inc\coreclrhost.h" Link="Templates\coreclrhost.h" />
```
Comment (am11) [inline src/tasks/AppleAppBuilder/AppleAppBuilder.csproj:12]:
```suggestion
    <EmbeddedResource Include="$(CoreClrProjectRoot)hosts\inc\coreclrhost.h" Link="Templates\coreclrhost.h" />
```

File: src/tasks/AppleAppBuilder/AppleAppBuilder.cs [hunk 3]
```diff
@@ -188,42 +188,47 @@ public string TargetOS
...
             if (!string.IsNullOrEmpty(DiagnosticPorts))
-                throw new ArgumentException($"Property \"{nameof(DiagnosticPorts)}\" is not supported with NativeAOT runtime and will be ignored.");
+                throw new ArgumentException($"Property \"{nameof(DiagnosticPorts)}\" is not supported with {Runtime} runtime and will be ignored.");
 
             if (EnableRuntimeLogging)
-                throw new ArgumentException($"Property \"{nameof(EnableRuntimeLogging)}\" is not supported with NativeAOT runtime and will be ignored.");
+                throw new ArgumentException($"Property \"{nameof(EnableRuntimeLogging)}\" is not supported with {Runtime} runtime and will be ignored.");
         }
-        else
+        else if (Runtime == "MonoVM" || Runtime == "")
```
Comment (am11) [inline src/tasks/AppleAppBuilder/AppleAppBuilder.cs:217]:
```suggestion
        else if (Runtime == "MonoVM" || string.IsNullOrEmpty(Runtime))
```

File: src/native/libs/System.Security.Cryptography.Native.Apple/entrypoints.c [hunk 4]
```diff
@@ -19,17 +18,25 @@
...
+#include "pal_x509_macos.h"
+#else
+#include "pal_keychain_ios.h"
+#include "pal_x509_ios.h"
+#endif
 
 static const Entry s_cryptoAppleNative[] =
 {
+#ifdef TARGET_OSX
     DllImportEntry(AppleCryptoNative_AesGcmEncrypt)
```
Comment (am11) [inline src/native/libs/System.Security.Cryptography.Native.Apple/entrypoints.c:35]: nit: maybe we can group these together and put them under a same ifdef?

File: src/coreclr/hosts/inc/coreclrhost.h (hunk 1 above)
Comment (am11) [reply src/coreclr/hosts/inc/coreclrhost.h:24]: Ah ok, maybe this header could use the same preamble as the other shared headers:

```h
#ifdef __cplusplus
extern "C"
{
#endif // __cplusplus

... code ...

#ifdef __cplusplus
}
#endif // __cplusplus
```

File: src/native/libs/System.Security.Cryptography.Native.Apple/entrypoints.c (hunk 4 above)
Comment (filipnavara) [reply src/native/libs/System.Security.Cryptography.Native.Apple/entrypoints.c:35]: I guess so... I originally thought they had to be sorted but then I found out they are not sorted in System.Native.

File: src/coreclr/hosts/inc/coreclrhost.h (hunk 1 above)
Comment (filipnavara) [reply src/coreclr/hosts/inc/coreclrhost.h:24]: I have no personal preference for either style. They are nearly the same in terms of changed lines.

**Comment by teo-tsirpanis** [comment]
@filipnavara do you plan to continue working on this?

**Comment by filipnavara** [comment]
Yes, it needs a merge/rebase to resolve conflicts.

(I am sick at the moment, so it may have to wait a few days)

**Comment by teo-tsirpanis** [comment]
OK, reopening, get well soon!

**Pull request closed without merging**